core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java

```diff
@@ -61,6 +61,15 @@ public ReplicationRequest(ActionRequest request) {
     }
 
     /**
+     * Creates a new request with resolved shard id
+     */
+    public ReplicationRequest(ActionRequest request, ShardId shardId) {
+        super(request);
+        this.index = shardId.getIndex();
+        this.internalShardId = shardId;
```

Comment:
can we rename internalShardId to resolvedShardId ? also add some java docs on how it should be used. I think that , as a separate change we can move the index field to another class DocReplicationRequest that is a subclass of this and a parent of IndexRequest and friends. It will be cleaner. No need to do it now.